src/components/Autocomplete/components/ComboBox/ComboBox.tsx

```diff
@@ -53,427 +41,374 @@ export interface ComboBoxProps {
 
-  componentDidMount() {
-    const {options, actionsBefore, actionsAfter} = this.props;
-    const comboBoxId = this.getComboBoxId();
-    let navigableOptions: (OptionDescriptor | ActionListItemDescriptor)[] = [];
-
-    if (actionsBefore) {
-      navigableOptions = navigableOptions.concat(actionsBefore);
-    }
-    if (options) {
-      navigableOptions = navigableOptions.concat(options);
-    }
-    if (actionsAfter) {
-      navigableOptions = navigableOptions.concat(actionsAfter);
-    }
-    navigableOptions = assignOptionIds(navigableOptions, comboBoxId);
+  const visuallyUpdateSelectedOption = useCallback(
+    (
+      newOption: OptionDescriptor | ActionListItemDescriptor,
+      oldOption: OptionDescriptor | ActionListItemDescriptor | undefined,
+    ) => {
+      if (oldOption) {
+        oldOption.active = false;
+      }
+      if (newOption) {
+        newOption.active = true;
+      }
+    },
+    [],
+  );
 
-    this.setState({
-      navigableOptions,
-    });
-  }
+  const resetVisuallySelectedOptions = useCallback(() => {
+    setSelectedOption(undefined);
+    setSelectedIndex(-1);
+    navigableOptions &&
+      navigableOptions.forEach(
+        (option: OptionDescriptor | ActionListItemDescriptor) => {
+          option.active = false;
+        },
+      );
+  }, [navigableOptions]);
 
-  componentDidUpdate(_: ComboBoxProps, prevState: State) {
-    const {contentBefore, contentAfter, emptyState} = this.props;
-    const {navigableOptions, popoverWasActive} = this.state;
+  const selectOptionAtIndex = useCallback(
+    (newOptionIndex: number) => {
+      if (!navigableOptions || navigableOptions.length === 0) {
+        return;
+      }
 
-    const optionsChanged =
-      navigableOptions &&
-      prevState.navigableOptions &&
-      !optionsAreEqual(navigableOptions, prevState.navigableOptions);
+      const newSelectedOption = navigableOptions[newOptionIndex];
 
-    if (optionsChanged) {
-      this.updateIndexOfSelectedOption(navigableOptions);
-    }
+      visuallyUpdateSelectedOption(newSelectedOption, selectedOption);
 
-    if (
-      navigableOptions &&
-      navigableOptions.length === 0 &&
-      !contentBefore &&
-      !contentAfter &&
-      !emptyState
-    ) {
-      // eslint-disable-next-line react/no-did-update-set-state
-      this.setState({popoverActive: false});
-    } else if (
-      popoverWasActive &&
-      navigableOptions &&
-      navigableOptions.length !== 0
-    ) {
-      // eslint-disable-next-line react/no-did-update-set-state
-      this.setState({popoverActive: true});
-    }
-  }
+      setSelectedOption(newSelectedOption);
+      setSelectedIndex(newOptionIndex);
+    },
+    [navigableOptions, selectedOption, visuallyUpdateSelectedOption],
+  );
 
-  getComboBoxId(): string {
-    if (this.state && this.state.comboBoxId) {
-      return this.state.comboBoxId;
+  const selectNextOption = useCallback(() => {
+    if (!navigableOptions || navigableOptions.length === 0) {
```

Comment:
it looks like navigableOptions is always an array so I don't think you need the first check as it will always be truthy

```suggestion
    if (navigableOptions.length === 0) {
```


You perform this check in a bunch of places, I think they can all be simplified.